### PR TITLE
Add fixSSH fallback

### PR DESCRIPTION
**Pull request by mjudeikis**
### Which issue this PR addresses:

Fixes ICM 250423613
workarounds partially https://bugzilla.redhat.com/show_bug.cgi?id=1882490

### What this PR does / why we need it:

If customer deleted master it will get re-created with different nic name (see bz for more info). If this happens we have following things:
1. Master is not in the LB anymore (not the scope of this PR)
2. fixSSH breaks with unknown NIC name (this pr)

### Test plan for issue:

Unit tests

### Is there any documentation that needs to be updated for this PR?

No

### CODE REVIEW

File: pkg/cluster/fixssh.go [hunk 1]
```diff
@@ -49,11 +49,26 @@ func (m *manager) fixSSH(ctx context.Context) error {
 	}
 
 	for i := 0; i < 3; i++ {
-		nicName := fmt.Sprintf("%s-master%d-nic", infraID, i)
-
-		nic, err := m.interfaces.Get(ctx, resourceGroup, nicName, "")
+		// NIC names might be different if customer re-created master nodes
+		// see https://bugzilla.redhat.com/show_bug.cgi?id=188249 for more details
```
Comment (rolandmkunkel) [inline pkg/cluster/fixssh.go:53]: should be https://bugzilla.redhat.com/show_bug.cgi?id=1882490

File: pkg/cluster/fixssh.go [hunk 2]
```diff
@@ -49,11 +49,26 @@ func (m *manager) fixSSH(ctx context.Context) error {
...
+		nicNameMachineAPI := fmt.Sprintf("%s-master-%d-nic", infraID, i)
+
+		var iErr, err error
+		var nic mgmtnetwork.Interface
+		nicName := nicNameInstaller
+
+		nic, err = m.interfaces.Get(ctx, resourceGroup, nicName, "")
 		if err != nil {
-			return err
+			iErr = err // preserve original error
```
Comment (rolandmkunkel) [inline pkg/cluster/fixssh.go:65]: I would prefer to instead nest the error so that the second one is not swallowed

File: pkg/cluster/fixssh.go [hunk 3]
```diff
@@ -49,11 +49,26 @@ func (m *manager) fixSSH(ctx context.Context) error {
...
+		// NIC names might be different if customer re-created master nodes
+		// see https://bugzilla.redhat.com/show_bug.cgi?id=188249 for more details
+		// installer naming  - <foo>-master{0,1,2}-nic
+		// machineAPI naming - <foo>-master-{0,1,2}-nic
+		nicNameInstaller := fmt.Sprintf("%s-master%d-nic", infraID, i)
+		nicNameMachineAPI := fmt.Sprintf("%s-master-%d-nic", infraID, i)
+
+		var iErr, err error
+		var nic mgmtnetwork.Interface
+		nicName := nicNameInstaller
```
Comment (rolandmkunkel) [inline pkg/cluster/fixssh.go:61]: this variable is not really necessary

File: pkg/cluster/fixssh.go [hunk 4]
```diff
@@ -49,11 +49,26 @@ func (m *manager) fixSSH(ctx context.Context) error {
...
+
+		nic, err = m.interfaces.Get(ctx, resourceGroup, nicName, "")
 		if err != nil {
-			return err
+			iErr = err // preserve original error
+			nicName = nicNameMachineAPI
+			m.log.Warnf("fallback to check MachineAPI Nic name format for %s", nicName)
+			nic, err = m.interfaces.Get(ctx, resourceGroup, nicName, "")
+			if err != nil {
+				return iErr
```
Comment (troy0820) [inline pkg/cluster/fixssh.go:70]: This error will be the error from the original preserved error and not the error from the failure of the nicNameMachineAPI interface get error.  _(It will show the wrong nic in the error)_

Also, it would read easier if instead of using `nicName` in both of the functions to use `nicNameInstaller`,`nicNameMachineAPI` where you are calling the interfaces.Get

**Review by troy0820 — CHANGES_REQUESTED** [review]
This looks good, just a couple of comments, but running into this yesterday was not fun.  Thanks for looking into this @mjudeikis 👍🏽

File: pkg/cluster/fixssh.go (hunk 4 above)
Comment (nilsanderselde) [reply pkg/cluster/fixssh.go:70]: Regarding the returned error, we'd only be in this block if `err`, and therefore `iErr` (sounds like an Apple product :), is not `nil`. Therefore, I think it's appropriate to return the original error (`iErr`).

As far as the use of `nicName`, this also appears correct to me, since it is used below this block as well, on lines 77 and 78. Setting `nicName := nicNameInstaller` as the default and then only overriding it if it can't find it is highly readable, as you can see at a glance what is the chosen value at any given point.

tl;dr: I think this is correct as is.

**Review by nilsanderselde — CHANGES_REQUESTED** [review]
Looks good, aside from the missing final digit in the BZ URL

File: pkg/cluster/fixssh.go (hunk 4 above)
Comment (mjudeikis) [reply pkg/cluster/fixssh.go:70]: Basically what Nils said :) 
We care about original error and that this is happen in the first place. Second error return would mascarade the real failure. I gonna log fallback error but not return it.

File: pkg/cluster/fixssh.go (hunk 2 above)
Comment (mjudeikis) [reply pkg/cluster/fixssh.go:65]: We dont nest errors in this code :/

File: pkg/cluster/fixssh.go [hunk 5]
```diff
@@ -49,11 +49,27 @@ func (m *manager) fixSSH(ctx context.Context) error {
...
+
+		nic, err = m.interfaces.Get(ctx, resourceGroup, nicName, "")
 		if err != nil {
-			return err
+			iErr = err // preserve original error
+			nicName = nicNameMachineAPI
+			m.log.Warnf("fallback to check MachineAPI Nic name format for %s", nicName)
+			nic, err = m.interfaces.Get(ctx, resourceGroup, nicName, "")
+			if err != nil {
+				m.log.Warnf("fallback failed with errir %s", err)
```
Comment (troy0820) [inline pkg/cluster/fixssh.go:70]:
```suggestion
				m.log.Warnf("fallback failed with err %s", err)
```
typo?

**Review by troy0820 — COMMENTED** [review]
nit: typo?

File: pkg/cluster/fixssh.go (hunk 4 above)
Comment (troy0820) [reply pkg/cluster/fixssh.go:71]: That makes sense.  If it fails both, returning the first error is returned and logging fallback 👍🏽 .

File: pkg/cluster/fixssh.go (hunk 5 above)
Comment (mjudeikis) [reply pkg/cluster/fixssh.go]: github copilot :D

**Review by nilsanderselde — APPROVED** [review]
